system/jlib/jstats.cpp

```diff
@@ -1988,19 +2087,71 @@ class CStatisticCollection : public CInterfaceOf<IStatisticCollection>
                 cur.visit(visitor);
         }
     }
-
     virtual void visitChildren(IStatisticVisitor & visitor) const
     {
         for (auto const & cur : children)
             cur.visit(visitor);
     }
+    virtual void refreshAggregates(const StatisticsMapping & mapping, std::function<void(const char * scope, StatisticScopeType sst, StatisticKind kind, stat_type value)> & fWhenAggregateUpdated) override
+    {
+        if (isDirty)
+        {
+            CRuntimeStatisticCollection totals(mapping);
+            refreshAggregates(totals, fWhenAggregateUpdated);
+        }
+    }
+    virtual stat_type aggregateStatistic(StatisticKind kind) const override
+    {
+        stat_type sum;
+        if (!getStatisticSum(kind, sum)) // get sum of statistics at this level
+        {
+            // if no stats at this level, then get sum of stats from children
+            for (auto & child : children)
+                sum += child.aggregateStatistic(kind);
+        }
+        return sum;
+    }
+    virtual void recordStats(const StatisticsMapping & mapping, IStatisticCollection * sourceStatsCollection, std::initializer_list<const StatsScopeId> path) override
+    {
+        CStatisticCollection * curSrcCollection = static_cast<CStatisticCollection *>(sourceStatsCollection);
+        const StatsScopeId * scopeItem = path.begin();
 
+        // Locate the collection with the stats and make curSrcCollection point to that
+        if (!curSrcCollection || curSrcCollection->queryScopeId().compare(*scopeItem)!=0)
+            return; // Required path doesn't exist in source collection so nothing more to do here
+        ++scopeItem;
+        while (scopeItem!=path.end())
+        {
+            curSrcCollection = curSrcCollection->children.find(scopeItem);
+            if (!curSrcCollection)
+                return; // Required path doesn't exist in source collection so nothing more to do here
+            ++scopeItem;
+        }
```

Comment:
see comments in workunit.cpp:2700, re. short-cutting the re-discovery of the correct IStatisticCollection, and using getFullScope.
And moving recordStats out of CStatsCollection + using existing methods to query kind/values